sputnikdao-factory2/src/lib.rs

```diff
@@ -109,30 +80,115 @@ impl SputnikDAOFactory {
             false
         }
     }
+
+    /// Tries to upgrade given account created by this factory to the latest code.
+    pub fn upgrade(&self, account_id: AccountId) {
+        self.assert_owner();
+        assert!(
+            self.daos.contains(&account_id),
+            "Must be contract created by factory"
+        );
+        self.factory_manager
+            .update_contract(account_id, self.get_latest_code_hash(), "update");
+    }
+
+    pub fn get_dao_list(&self) -> Vec<AccountId> {
+        self.daos.to_vec()
+    }
+
+    /// Get number of created DAOs.
+    pub fn get_number_daos(&self) -> u64 {
+        self.daos.len()
+    }
+
+    /// Get daos in paginated view.
+    pub fn get_daos(&self, from_index: u64, limit: u64) -> Vec<AccountId> {
+        let elements = self.daos.as_vector();
+        (from_index..std::cmp::min(from_index + limit, elements.len()))
+            .filter_map(|index| elements.get(index))
+            .collect()
+    }
+
+    pub fn get_owner(&self) -> AccountId {
+        AccountId::new_unchecked(
+            String::from_utf8(
+                env::storage_read(OWNER_KEY)
+                    .unwrap_or(env::current_account_id().as_bytes().to_vec()),
+            )
+            .expect("INTERNAL_FAIL"),
+        )
+    }
+
+    pub fn get_latest_code_hash(&self) -> Base58CryptoHash {
+        slice_to_hash(&env::storage_read(LATEST_CODE_HASH_KEY).expect("Must have code hash"))
+    }
+
+    /// Returns non serialized code by given code hash.
+    pub fn get_code(&self, code_hash: Base58CryptoHash) {
+        self.factory_manager.get_code(code_hash);
+    }
+
+    fn assert_owner(&self) {
+        assert_eq!(
+            self.get_owner(),
+            env::predecessor_account_id(),
+            "Must be owner"
+        );
+    }
+}
+
+pub fn slice_to_hash(hash: &[u8]) -> Base58CryptoHash {
+    let mut result: CryptoHash = [0; 32];
+    result.copy_from_slice(&hash);
+    Base58CryptoHash::from(result)
+}
+
+/// Store new contract. Non serialized argument is the contract.
+/// Returns base58 of the hash of the contract.
+#[no_mangle]
+pub extern "C" fn store() {
```

Comment:
Feels like this `store` function should also check for the owner, or else we might have random folks uploading blobs?
`self.assert_owner();` or something similar